### PR TITLE
[TheiaEuk] TheiaEuk ONT Workflow

### LINKED ISSUE #625: [New workflow] TheiaEuk_ONT
:cool:

### PR DESCRIPTION
**Pull request by Michal-Babins**
This PR closes #625 

🗑️ This dev branch should <NOT> be deleted after merging to main.

## :brain: Summary

Creates a TheiaEuk workflow for ONT reads

## :zap: Impacted Workflows/Tasks

- New workflow: wf_theiaeuk_ont
- Modified: wf_merlin_magic (to accommodate eukaryotic-specific analyses)
- Modified: task_snippy_variants and task_snippy_gene_query (to work with ONT data)
- Modified: wf_read_QC_trim_ont for theiaeuk specific workflows

This PR may lead to different results in pre-existing outputs: **No**

This PR uses an element that could cause duplicate runs to have different results: **No**

### :gear: Algorithm

- Included theiaeuk specific qc trimming happy path to wf_read_QC_trim_ont
- Adapted Snippy for variant calling with ONT data 
- Included Candida Aruis ONT specific analyses in the Merlin Magic workflow

### ➡️ Inputs

The following inputs are mandatory for theiaeuk_ont
```
File read1
String samplename
```
The following inputs are optional for theiaeuk_ont
```
Int genome_length
String? assembler
String? assembler_options
Int dragonflye_cpu 
Int dragonflye_memory
Int dragonflye_disk_size
String medaka_model 
File gambit_db_genomes 
File gambit_db_signatures
```

### ⬅️ Outputs

```
# Version Capture
    String theiaeuk_ont_version = version_capture.phb_version
    String theiaeuk_ont_analysis_date = version_capture.date
    # Read QC outputs
    File read1_clean = read_qc.read1_clean
    String? nanoq_version = read_qc.nanoq_version
    Int est_genome_length = read_qc.est_genome_length
    # Assembly outputs
    File assembly_fasta = dragonflye.assembly_fasta
    File contigs_gfa = dragonflye.contigs_gfa
    String dragonflye_version = dragonflye.dragonflye_version
    # Read QC - nanoplot raw outputs
    File? nanoplot_html_raw = nanoplot_raw.nanoplot_html
    File? nanoplot_tsv_raw = nanoplot_raw.nanoplot_tsv
    Int? nanoplot_num_reads_raw1 = nanoplot_raw.num_reads
    Float? nanoplot_r1_median_readlength_raw = nanoplot_raw.median_readlength
    Float? nanoplot_r1_mean_readlength_raw = nanoplot_raw.mean_readlength
    Float? nanoplot_r1_stdev_readlength_raw = nanoplot_raw.stdev_readlength
    Float? nanoplot_r1_n50_raw = nanoplot_raw.n50
    Float? nanoplot_r1_mean_q_raw = nanoplot_raw.mean_q
    Float? nanoplot_r1_median_q_raw = nanoplot_raw.median_q
    Float? nanoplot_r1_est_coverage_raw = nanoplot_raw.est_coverage
    # Read QC - nanoplot clean outputs
    File? nanoplot_html_clean = nanoplot_clean.nanoplot_html
    File? nanoplot_tsv_clean = nanoplot_clean.nanoplot_tsv
    Int? nanoplot_num_reads_clean1 = nanoplot_clean.num_reads
    Float? nanoplot_r1_median_readlength_clean = nanoplot_clean.median_readlength
    Float? nanoplot_r1_mean_readlength_clean = nanoplot_clean.mean_readlength
    Float? nanoplot_r1_stdev_readlength_clean = nanoplot_clean.stdev_readlength
    Float? nanoplot_r1_n50_clean = nanoplot_clean.n50
    Float? nanoplot_r1_mean_q_clean = nanoplot_clean.mean_q
    Float? nanoplot_r1_median_q_clean = nanoplot_clean.median_q
    Float? nanoplot_r1_est_coverage_clean = nanoplot_clean.est_coverage
    # Read QC - nanoplot general outputs
    String? nanoplot_version = nanoplot_raw.nanoplot_version
    String? nanoplot_docker = nanoplot_raw.nanoplot_docker
    # Assembly QC - quast outputs
    File? quast_report = quast.quast_report
    String? quast_version = quast.version
    Int? assembly_length = quast.genome_length
    Int? number_contigs = quast.number_contigs
    Int? n50_value = quast.n50_value
    Float? quast_gc_percent = quast.gc_percent
    # Assembly QC - nanoplot outputs
    Float? est_coverage_raw = nanoplot_raw.est_coverage
    Float? est_coverage_clean = nanoplot_clean.est_coverage
    # Assembly QC - busco outputs
    String? busco_version = busco.busco_version
    String? busco_docker = busco.busco_docker
    String? busco_database = busco.busco_database
    String? busco_results = busco.busco_results
    File? busco_report = busco.busco_report
    # Gambit outputs
    File gambit_report_file = gambit.gambit_report_file
    File gambit_closest_genomes_file = gambit.gambit_closest_genomes_file
    String gambit_predicted_taxon = gambit.gambit_predicted_taxon
    String gambit_predicted_taxon_rank = gambit.gambit_predicted_taxon_rank
    String gambit_next_taxon = gambit.gambit_next_taxon
    String gambit_next_taxon_rank = gambit.gambit_next_taxon_rank
    String gambit_version = gambit.gambit_version
    String gambit_db_version = gambit.gambit_db_version
    String merlin_tag = gambit.merlin_tag
    String gambit_docker = gambit.gambit_docker
    # C. auris specific outputs for cladetyper
    String? clade_type = merlin_magic.clade_type
    String? cladetyper_analysis_date = merlin_magic.cladetyper_analysis_date
    String? cladetyper_version = merlin_magic.cladetyper_version
    String? cladetyper_docker_image = merlin_magic.cladetyper_docker_image
    String? cladetype_annotated_ref = merlin_magic.cladetype_annotated_ref
    # Snippy variants outputs
    String? snippy_variants_version = merlin_magic.snippy_variants_version
    String? snippy_variants_query = merlin_magic.snippy_variants_query
    String? snippy_variants_query_check = merlin_magic.snippy_variants_query_check
    String? snippy_variants_hits = merlin_magic.snippy_variants_hits
    String? snippy_variants_gene_query_results = merlin_magic.snippy_variants_gene_query_results
    String? snippy_variants_outdir_tarball = merlin_magic.snippy_variants_outdir_tarball
    String? snippy_variants_results = merlin_magic.snippy_variants_results
    String? snippy_variants_bam = merlin_magic.snippy_variants_bam
    String? snippy_variants_bai = merlin_magic.snippy_variants_bai
    String? snippy_variants_summary = merlin_magic.snippy_variants_summary
    String? snippy_variants_num_reads_aligned = merlin_magic.snippy_variants_num_reads_aligned
    String? snippy_variants_coverage_tsv = merlin_magic.snippy_variants_coverage_tsv
    String? snippy_variants_num_variants = merlin_magic.snippy_variants_num_variants
    String? snippy_variants_percent_ref_coverage = merlin_magic.snippy_variants_percent_ref_coverage
```

## :microscope: Final Developer Checklist

- [x] The workflow/task has been tested and results, including file contents, are as anticipated
- [x] The CI/CD has been adjusted and tests are passing (Theiagen developers)
- [x] Code changes follow the [style guide](https://theiagen.notion.site/Style-Guide-WDL-Workflow-Development-51b66a47dde54c798f35d673fff80249)
- [x] Documentation and/or workflow diagrams have been updated if applicable (Theiagen developers only)

## 🎯 Reviewer Checklist

- [ ] All changed results have been confirmed
- [ ] You have tested the PR appropriately (see the [testing guide](https://theiagen.notion.site/PR-Testing-Guide-Determining-Appropriate-Levels-of-Testing-4764e98a6aeb460185039c0896714590) for more information)
- [x] All code adheres to the [style guide](https://theiagen.notion.site/Style-Guide-WDL-Workflow-Development-51b66a47dde54c798f35d673fff80249)
- [x] MD5 sums have been updated
- [x] The PR author has addressed all comments
- [x] The documentation has been updated

### CODE REVIEW

**Comment by michellescribner** [comment]
I've launched some initial function tests [here](https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Scribner_Sandbox/job_history/9654c299-3208-4e97-8758-fb3219e6adbd) using ONT data from SRA for fungal pathogens and intend to explore how closely assembly lengths align with expected values. This is not intended to be comprehensive testing as I do not have basecalling metadata for these reads.

**Comment by cimendes** [comment]
image
Seeing this on the snippy_variants_num_variants, which doesn't match the expected results, but I haven't figured out what is causing it! 

For comparison, here's the same column for ILMN:
image

**Comment by cimendes** [comment]
To alingn with TheiaEuk_Illumina_PE, clade_type should be renamed to cladetyper_clade

**Comment by cimendes** [comment]
The latest version of this workflow ( commit #[893ab9c](https://github.com/theiagen/public_health_bioinformatics/pull/644/commits/893ab9cd6daa1bf80134d54130f7b157247fa69c)) was end-to-end tested at https://app.terra.bio/#workspaces/theiagen-validations/ines_sandbox/submission_history/43bb8023-dd61-41b4-9a9b-7f8abaf6c34f. After comparison with TheiaEuk_Illumina_PE_PHB version v3.0.1 on a set of 16 C. auris samples (where TheiaValidate was used, run available [here](https://app.terra.bio/#workspaces/theiagen-validations/ines_sandbox/submission_history/75ee514a-05de-424a-adc9-9e4076ec46cc)), the only differences detected in the genomic characterization modules were in the snippy hits results, and only for 3/16 samples.  This is expected as variant calling for ONT is being performed on the assemblies that are fed to snippy, and not on the read level, like how it happens in ILMN data.

The improvement on variant calling could be addressed in Phase II of this project. For Phase I, this workflow provided consistent, high-quality assemblies for C. auris, which was the ultimate goal. A thorough assessment was carried out and available in our Drive. 

TL;DR: I consider this PR ready to be merged, after code review, which was not carried out by me. The documentation also needs to be reviewed, which I will do after it's completed.

File: workflows/utilities/wf_merlin_magic.wdl [hunk 1]
```diff
@@ -224,9 +225,9 @@ workflow merlin_magic {
     String tbp_parser_output_seq_method_type = "WGS"
     String? tbp_parser_operator
     Int? tbp_parser_min_depth
-    Float? tbp_parser_min_frequency
+    Int? tbp_parser_min_frequency
```
Comment (sage-wright) [inline workflows/utilities/wf_merlin_magic.wdl:228]: these should be floats

File: workflows/utilities/wf_read_QC_trim_ont.wdl [hunk 2]
```diff
@@ -195,7 +170,29 @@ workflow read_QC_trim_ont {
...
+            memory = kraken_memory,
+            cpu = kraken_cpu
+        }
+        call kraken2.kraken2_parse_classified as kraken2_recalculate_abundances {
+          input:
+            samplename = samplename,
+            kraken2_report = kraken2_se.kraken2_report,
+            kraken2_classified_report = kraken2_se.kraken2_classified_report
+        } 
+      } if ((call_kraken) && ! defined(kraken_db)) {
```
Comment (sage-wright) [inline workflows/utilities/wf_read_QC_trim_ont.wdl:193]: pls fix this indentation and remove reference to tiptoft

File: workflows/theiaeuk/wf_theiaeuk_ont.wdl [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+  # call merlin magic for cladetyper and AMR search, snippy variants
+   call merlin_magic_workflow.merlin_magic {
+    input:
+      samplename = samplename,
+      merlin_tag = gambit.merlin_tag,
+      assembly = flye_denovo.assembly_fasta,
+      assembly_only = true,
+      ont_data = true,
+      theiaeuk = true
+  }
```
Comment (sage-wright) [inline workflows/theiaeuk/wf_theiaeuk_ont.wdl:84]: your diagram advertises the qc check module but it's not implemented; one of those should be amended

Comment (Michal-Babins) [reply workflows/theiaeuk/wf_theiaeuk_ont.wdl:84]: @cimendes these weren't part of phase one implementation were they?

File: docs/assets/tables/all_inputs.tsv [hunk 4]
```diff
@@ -290,12 +292,12 @@ biosample_submit_tsv_ftp_upload	cpu	Int	Number of CPUs to allocate to the task	2
...
-busco	memory	Int	Amount of memory/RAM (in GB) to allocate to the task	8	Optional	TheiaProk_FASTA, TheiaProk_ONT, TheiaProk_Illumina_PE, TheiaProk_Illumina_SE	
+busco	cpu	Int	Number of CPUs to allocate to the task	2	Optional	TheiaEuk_Illumina_PE, TheiaProk_FASTA, TheiaProk_ONT, TheiaProk_Illumina_PE, TheiaProk_Illumina_SE, TheiaEuk_ONT
+busco	disk_size	Int	Amount of storage (in GB) to allocate to the task	100	Optional	TheiaEuk_Illumina_PE, TheiaProk_FASTA, TheiaProk_ONT, TheiaProk_Illumina_PE, TheiaProk_Illumina_SE, TheiaEuk_ONT	
+busco	docker	String	The Docker container to use for the task	us-docker.pkg.dev/general-theiagen/ezlabgva/busco:v5.3.2_cv1	Optional	TheiaEuk_Illumina_PE, TheiaEuk_ONT
+busco	docker	String	The Docker container to use for the task	us-docker.pkg.dev/general-theiagen/ezlabgva/busco:v5.7.1_cv1	Optional	TheiaProk_FASTA, TheiaProk_ONT, TheiaProk_Illumina_PE, TheiaProk_Illumina_SE, TheiaEuk_ONT	
+busco	eukaryote	Boolean	Assesses eukaryotic organisms, rather than prokaryotic organisms	FALSE	Optional	TheiaProk_FASTA, TheiaProk_ONT, TheiaProk_Illumina_PE, TheiaProk_Illumina_SE, TheiaEuk_ONT
```
Comment (sage-wright) [inline docs/assets/tables/all_inputs.tsv:299]: busco inputs are incorrect.

- the only busco task inputs are busco cpu and busco disk_size.

- busco eukaryote does not exist for theiaeuk ont

- the other busco inputs are at the workflow level:
  - theiaeuk_ont busco_docker_image
  - theiaeuk_ont busco_memory 
  
these should be corrected

File: docs/assets/tables/all_inputs.tsv [hunk 5]
```diff
@@ -1404,22 +1406,22 @@ read_QC_trim	trim_quality_trim_score	Int	Specifies the average quality of bases
...
+read_QC_trim_ont	kraken_memory	Int	Amount of memory/RAM (in GB) to allocate to the task	8	Optional	Freyja_FASTQ (ONT), TheiaEuk_ONT
+read_QC_trim_ont	max_length	Int	Internal component, do not modify		Optional	Freyja_FASTQ (ONT), TheiaEuk_ONT
+read_QC_trim_ont	kraken2_recalculate_abundances_cpu	Int	Number of CPUs to allocate to the task	4	Optional	Freyja_FASTQ (ONT), TheiaEuk_ONT
+read_QC_trim_ont	kraken2_recalculate_abundances_disk_size	Int	Amount of storage (in GB) to allocate to the task	100	Optional	Freyja_FASTQ (ONT), TheiaEuk_ONT
+read_QC_trim_ont	kraken2_recalculate_abundances_docker	Int	The Docker container to use for the task	"us-docker.pkg.dev/general-theiagen/theiagen/terra-tools:2023-08-28-v4"	Optional	Freyja_FASTQ (ONT), TheiaEuk_ONT
+read_QC_trim_ont	kraken2_recalculate_abundances_memory	Int	Amount of memory/RAM (in GB) to allocate to the task	8	Optional	Freyja_FASTQ (ONT), TheiaEuk_ONT
+read_QC_trim_ont	min_length	Int	Internal component, do not modify		Optional	Freyja_FASTQ (ONT), TheiaEuk_ONT
+read_QC_trim_ont	run_prefix	String	Internal component, do not modify		Optional	Freyja_FASTQ (ONT), TheiaEuk_ONT
```
Comment (sage-wright) [inline docs/assets/tables/all_inputs.tsv:1422]: this is not an option for theiaeuk

File: docs/assets/tables/all_inputs.tsv [hunk 6]
```diff
@@ -1404,22 +1406,22 @@ read_QC_trim	trim_quality_trim_score	Int	Specifies the average quality of bases
...
+read_QC_trim_ont	kraken_disk_size	Int	Amount of storage (in GB) to allocate to the task	100	Optional	Freyja_FASTQ (ONT), TheiaEuk_ONT
+read_QC_trim_ont	kraken_memory	Int	Amount of memory/RAM (in GB) to allocate to the task	8	Optional	Freyja_FASTQ (ONT), TheiaEuk_ONT
+read_QC_trim_ont	max_length	Int	Internal component, do not modify		Optional	Freyja_FASTQ (ONT), TheiaEuk_ONT
+read_QC_trim_ont	kraken2_recalculate_abundances_cpu	Int	Number of CPUs to allocate to the task	4	Optional	Freyja_FASTQ (ONT), TheiaEuk_ONT
+read_QC_trim_ont	kraken2_recalculate_abundances_disk_size	Int	Amount of storage (in GB) to allocate to the task	100	Optional	Freyja_FASTQ (ONT), TheiaEuk_ONT
+read_QC_trim_ont	kraken2_recalculate_abundances_docker	Int	The Docker container to use for the task	"us-docker.pkg.dev/general-theiagen/theiagen/terra-tools:2023-08-28-v4"	Optional	Freyja_FASTQ (ONT), TheiaEuk_ONT
+read_QC_trim_ont	kraken2_recalculate_abundances_memory	Int	Amount of memory/RAM (in GB) to allocate to the task	8	Optional	Freyja_FASTQ (ONT), TheiaEuk_ONT
+read_QC_trim_ont	min_length	Int	Internal component, do not modify		Optional	Freyja_FASTQ (ONT), TheiaEuk_ONT
```
Comment (sage-wright) [inline docs/assets/tables/all_inputs.tsv:1421]: you are missing rasusa inputs

**Comment by Michal-Babins** [comment]
Test with read screen:
https://app.terra.bio/#workspaces/theiagen-validations/ines_sandbox/submission_history/58d798dd-2574-4b63-bc9b-af810ee5e8a2

File: docs/assets/tables/all_inputs.tsv [hunk 7]
```diff
@@ -889,195 +891,195 @@ mercury_prep_n_batch	authors_sbt	File	Only for "mpox" submissions; a file that c
...
+merlin_magic	amr_search_disk_size	Int	Amount of storage (in GB) to allocate to the task	50	Optional	TheiaEuk_Illumina_PE, TheiaProk_FASTA, TheiaProk_ONT, TheiaProk_Illumina_PE, TheiaProk_Illumina_SE, TheiaEuk_ONT	
+merlin_magic	amr_search_docker_image	String	The Docker container to use for the task	us-docker.pkg.dev/general-theiagen/theiagen/amrsearch:0.2.1	Optional	TheiaEuk_Illumina_PE, TheiaProk_FASTA, TheiaProk_ONT, TheiaProk_Illumina_PE, TheiaProk_Illumina_SE, TheiaEuk_ONT
+merlin_magic	amr_search_memory	Int	Amount of memory/RAM (in GB) to allocate to the task	8	Optional	TheiaEuk_Illumina_PE, TheiaProk_FASTA, TheiaProk_ONT, TheiaProk_Illumina_PE, TheiaProk_Illumina_SE, TheiaEuk_ONT
+merlin_magic	assembly_only	Boolean	Internal component, do not modify		Optional	TheiaEuk_Illumina_PE, TheiaProk_ONT, TheiaProk_Illumina_PE, TheiaProk_Illumina_SE, TheiaEuk_ONT
+merlin_magic	call_poppunk	Boolean	If true, runs PopPUNK for GPSC cluster designation for S. pneumoniae	TRUE	Optional	TheiaProk_FASTA, TheiaProk_ONT, TheiaProk_Illumina_PE, TheiaProk_Illumina_SE, TheiaEuk_ONT
```
Comment (sage-wright) [inline docs/assets/tables/all_inputs.tsv:908]: so you went all out here which is fab but you also ended up adding theiaeuk_ont to both call_poppunk vars (and several others) that it shouldn't be  

like, the cladetyper_docker isn't a do not use for theiaeuk, but you have it listed that it is. you also have theiaeuk_ont listed for prokaryotic-only options

File: docs/common_text/amr_search_task.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
+??? task "`amr_search`: Antimicrobial resistance profiling"
+
```
[the remark this answers is not in the document]
Comment (Michal-Babins) [reply docs/common_text/amr_search_task.md:2]: Highlighted in the inputs where default is FALSE

**Review by sage-wright — APPROVED** [review]
⭐ good job

**Review by sage-wright — DISMISSED** [review]
🥇 for real this time

**Review by sage-wright — APPROVED** [review]
third time's the charm 🌟
